bchec/field.go

```diff
@@ -305,32 +305,11 @@ func (f *fieldVal) Normalize() *fieldVal {
 	// following determines if either or these conditions are true and does
 	// the final reduction in constant time.
 	//
-	// Note that the if/else statements here intentionally do the bitwise
-	// operators even when it won't change the value to ensure constant time
-	// between the branches.  Also note that 'm' will be zero when neither
-	// of the aforementioned conditions are true and the value will not be
-	// changed when 'm' is zero.
-	m = 1
-	if t9 == fieldMSBMask {
-		m &= 1
-	} else {
-		m &= 0
-	}
-	if t2&t3&t4&t5&t6&t7&t8 == fieldBaseMask {
-		m &= 1
-	} else {
-		m &= 0
-	}
-	if ((t0+977)>>fieldBase + t1 + 64) > fieldBaseMask {
-		m &= 1
-	} else {
-		m &= 0
-	}
-	if t9>>fieldMSBBits != 0 {
-		m |= 1
-	} else {
-		m |= 0
-	}
+	// Note that 'm' will be zero when neither of the aforementioned conditions
+	// are true and the value will not be changed when 'm' is zero.
+	m = isZeroUint32((t9 - fieldMSBMask) | (t2&t3&t4&t5&t6&t7&t8 - fieldBaseMask))
+	m &= lessThanUint32(fieldBaseMask, (t0+977)>>fieldBase+t1+64)
+	m |= notZeroUint32(t9 >> fieldMSBBits)
```

Comment:
For reference, this seems to get done more simply in libsecp which also has an implementation of the same 10x26 field limbing: https://github.com/bitcoin-core/secp256k1/blob/0c774d89e61808d883b8f2df74462421491d59bb/src/field_10x26_impl.h#L66-L67

I have to admit I don't understand this deeply enough to know how either works.

Reply:
(for what it's worth, this PR is at least clearly identical to the code being modified, so there is nothing to worry about in terms of changed behaviour)

Reply:
OK I see now this is identical with libsecp... the two different meanings of 'm' was confusing me.

The libsecp expression is still simpler, for example it takes advantage of the fact that `t9 >> fieldMSBBits` is surely either 0 or 1 at this point. You could do the same by making the final line to be `m |= t9 >> fieldMSBBits`.

And libsecp uses equality comparisons along with implicit bool -> int cast (i.e., in the assembly it calls `cmp` which just does a subtraction anyway but only uses the result to set flags, then `sete` casts the flag to an int). Unfortunately it looks like in go you might not have a simple bool->int cast so this is about as good as it gets?

Reply:
Yes, the lack of bool->int cast is an unfortunate choice in golang when you want constant time math. This is as good as it gets and follows the standard patterns in [crypto/subtle](https://golang.org/src/crypto/subtle/constant_time.go).

It would be great to incorporate simplifications and improvements based on the libsecp implementation, modulo lack of above casts and different language standards for constants-as-variables.